libyul/backends/wasm/polyfill/Memory.yul

```diff
@@ -61,9 +61,49 @@ function memoryguard(x:i64) -> y1, y2, y3, y4 {
 	y4 := x
 }
 
+// Fill `length` bytes starting from `ptr` with `value`.
 function memset(ptr:i32, value:i32, length:i32) {
 	for { let i:i32 := 0:i32 } i32.lt_u(i, length) { i := i32.add(i, 1:i32) }
 	{
 		i32.store8(i32.add(ptr, i), value)
 	}
-}
\ No newline at end of file
+}
+
+// Writes 256-bits from `pos`, but only set the bottom 160-bits.
+function mstore_address(pos:i32, a1, a2, a3, a4) {
+	a1, a2, a3 := u256_to_address(a1, a2, a3, a4)
+	i64.store(pos, 0:i64)
+	i32.store(i32.add(pos, 8:i32), 0:i32)
+	i32.store(i32.add(pos, 12:i32), bswap32(i32.wrap_i64(a1)))
+	i64.store(i32.add(pos, 16:i32), bswap64(a2))
+	i64.store(i32.add(pos, 24:i32), bswap64(a3))
+}
```

Comment:
The thing is that `mstore_internal` and `mstore_address` are doing exactly the same from the encoding point of view. 

Maybe it is a bit confusing that `mstore_address` is using `i32.store`. I just used this to clean the "upper" 32 bytes by just setting it to zero using `i32.store(i32.add(pos, 8:i32), 0:i32)` and set the "lower" 32 bytes by using `i32.store(i32.add(pos, 12:i32), bswap32(i32.wrap_i64(a1)))`.

Lets say we wan't to store  the 256 bit value `0x00112233|44556677|8899aabb|ccddeeff|00112233|44556677|8899aabb|ccddeeff` using `mstore_internal`, the result will be byte swapped, the result will finally look like `0x77665544|33221100|ffeeddcc|bbaa9988|77665544|33221100|ffeeddcc|bbaa9988`. Basically `mstore_address` is doing exactly the same, but ensures that only 160 bit will be set. It will fail if other bits where set. That means using `0x00112233|44556677|8899aabb|ccddeeff|00112233|44556677|8899aabb|ccddeeff` will result in a failure, but if we only set the correct address bits like e.g. `0x00000000|00000000|00000000|11223344|55667788|99aabbcc|ddeeff00|11223344`  the result will be 
`0x00000000|00000000|44332211|00000000|ccbbaa99|88776655|44332211|00ffeedd`.
Of course, setting `0x00000000|00000000|00000000|11223344|55667788|99aabbcc|ddeeff00|11223344`   using `mstore_internal` will return exactly the same result
`0x00000000|00000000|44332211|00000000|ccbbaa99|88776655|44332211|00ffeedd`.

Reply:
> if we only set the correct address bits like e.g. `0x00000000|00000000|00000000|11223344|55667788|99aabbcc|ddeeff00|11223344`  the result will be 
`0x00000000|00000000|44332211|00000000|ccbbaa99|88776655|44332211|00ffeedd`.

In that case either there's still something I don't understand or the code has a bug. This code:
```
	i32.store(i32.add(pos, 8:i32), 0:i32)
	i32.store(i32.add(pos, 12:i32), bswap32(i32.wrap_i64(a1)))
```
will store `0x00000000` at bytes 8-11 and `0x44332211` (the lowest 4 bytes of `a2`) at bytes 12-15. So we'll actually get

```
0x00000000|00000000|00000000|44332211|ccbbaa99|88776655|44332211|00ffeedd
```

Reply:
Good question @cameel! By answering it I got much more conscious about what is really happening here. 

My answer from last time was not entirely correct, I missed two important things:
- in-memory representation of little-endian -> in-memory representation of data in little-endian is exactly the other way around.
- use of hera's `debug.print64` to get the actual values in the last answer, that also switched the endianness.

If we want to store the following address
`0x[00000000|00000000]|[00000000|11223344]|[55667788|99aabbcc]|[ddeeff00|11223344]`. I used `[]` to group 64 bits.
```
a1 := 0x0000000011223344 (wasm, little-endian) -> 0x4433221100000000 (in memory)
a2 := 0x5566778899aabbcc (wasm, little-endian) -> 0xccbbaa9988776655 (in memory)
a3 := 0xddeeff0011223344 (wasm, little-endian) -> 0x4433221100ffeedd (in memory)
```
Lets say we have the following 32 bytes data in memory:
```
   0x[ffffffff|ffffffff]|[ffffffff|ffffffff]|[ffffffff|ffffffff]|[ffffffff|ffffffff]
```
Lets follow the execution:
```
   0x[ffffffff|ffffffff]|[ffffffff|ffffffff]|[ffffffff|ffffffff]|[ffffffff|ffffffff]

i64.store(0, 0:i64)
   0x[00000000|00000000]|[ffffffff|ffffffff]|[ffffffff|ffffffff]|[ffffffff|ffffffff]

i32.store(i32.add(pos, 8:i32), 0:i32)
   0x[00000000|00000000]|[00000000|ffffffff]|[ffffffff|ffffffff]|[ffffffff|ffffffff]

i32.store(i32.add(pos, 12:i32), bswap32(i32.wrap_i64(a1)))
   0x[00000000|00000000]|[00000000|11223344]|[ffffffff|ffffffff]|[ffffffff|ffffffff]

i64.store(i32.add(pos, 16:i32), bswap64(a2))
   0x[00000000|00000000]|[00000000|11223344]|[55667788|99aabbcc]|[ffffffff|ffffffff]

i64.store(i32.add(pos, 24:i32), bswap64(a3))
   0x[00000000|00000000]|[00000000|11223344]|[55667788|99aabbcc]|[ddeeff00|11223344]
```
That means the "physical" memory would look like this:
```
   0x[00000000|00000000]|[00000000|11223344]|[55667788|99aabbcc]|[ddeeff00|11223344]
```
Using `debug.print64` in hera on x86_64:
```
   0x[00000000|00000000]|[00000000|11223344]|[55667788|99aabbcc]|[ddeeff00|11223344]
                         |                 | |                 | |                 |
                                  a1                  a2                  a3
  debug.print64(a1)
    0x0000000044332211
  debug.print64(a2)
    0x5566778899aabbcc
  debug.print64(a3)
    0x00ffeedd44332211
```
That means that `debug.print64` swaps the bytes again:
```
  0x[00000000|00000000]|[44332211|00000000]|[ccbbaa99|88776655]|[44332211|00ffeedd]
```
That's why I wrote `0x00000000|00000000|44332211|00000000|ccbbaa99|88776655|44332211|00ffeedd`.

Sorry for the confusion. I hope all the stuff makes now much more sense.

Reply:
Thanks! Now I see where I got it wrong. I knew that `i64.store()` will store numbers in little endian but somehow I neglected the fact that this means that I should write them reversed when showing the content of memory and I ended up with wrong results for each function.

My initial example should have looked like this - both functions will produce the same memory representation:
```
                     | 0      | 4      | 8      | 12     | 16     | 20     | 24     | 28     |
mstore_internal(): 0x 00000000 00000000 00000000 01000000 02000000 03000000 04000000 05000000
mstore_address():  0x 00000000 00000000 00000000 01000000 02000000 03000000 04000000 05000000 
```

In that case the code is fine.